src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.TextWriter.cs

```diff
@@ -0,0 +1,496 @@
+        public async Task EndTextWriterValueScopeAsync()
+        {
+            if (!IsWritingJson)
+            {
+                this.bufferWriter.Write(this.DoubleQuote.Slice(0, 1).Span);
+                await this.textWriter.DisposeAsync();
+                await this.FlushIfBufferThresholdReachedAsync().ConfigureAwait(false);
+
+                CheckIfSeparatorNeeded();
+                CheckIfManualValueAtArrayStart();
+            }
+        }
+
+        /// <summary>
+        /// Whether the current TextWriter is writing JSON
+        /// </summary>
+        /// <returns></returns>
+        private bool IsWritingJson
+        {
+            get
+            {
+                return String.IsNullOrEmpty(this.currentTextWriterContentType) || this.currentTextWriterContentType.StartsWith(MimeConstants.MimeApplicationJson, StringComparison.Ordinal);
+            }
+        }
+
+        /// <summary>
+        /// Represents a TextWriter implementation for writing UTF-8 JSON data in an OData context.
+        /// </summary>
+        internal class ODataUtf8JsonTextWriter : TextWriter
+        {
+            ODataUtf8JsonWriter jsonWriter = null;
+            private char[] buffer;
+            int numOfCharsNotWrittenFromPreviousChunk = 0;
+
+            public ODataUtf8JsonTextWriter(ODataUtf8JsonWriter jsonWriter)
+            {
+                this.jsonWriter = jsonWriter;
+            }
+
+            public override Encoding Encoding => throw new NotImplementedException();
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream synchronously.
+            /// </summary>
+            public override void Flush()
+            {
+                this.jsonWriter.Flush();
+            }
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream asynchronously.
+            /// </summary>
+            /// <param name="cancellationToken">A cancellation token to observe while waiting for the flush operation to complete.</param>
+            /// <returns>A task representing the asynchronous flush operation.</returns>
+            public override async Task FlushAsync()
+            {
+                await this.jsonWriter.FlushAsync();
+            }
+
+            /// <summary>
+            /// Disposes the object.
+            /// </summary>
+            /// <param name="disposing">True if called from Dispose; false if called form the finalizer.</param>
+            protected override void Dispose(bool disposing)
+            {
+                this.Flush();
+
+                // If there are unprocessed chars, encode and write them as the final block.
+                ReadOnlySpan<char> charsNotProcessedFromPreviousChunk = this.buffer.AsSpan().Slice(0, numOfCharsNotWrittenFromPreviousChunk);
+
+                if (!charsNotProcessedFromPreviousChunk.IsEmpty)
+                {
+                    if (!this.jsonWriter.IsWritingJson)

```

Comment:
Given that the content type won't change for this instance of the text writer, and given that that retrieiving `IsWritingJson` does some computation (checking whether a strings begins with a certain keyword), I think we can store this value inside a field in the `ODataUtf8JsonTextWriter` constructor and use that field in if statements instead.

Reply:
Can the `currentTextWriterContentType ` change after `ODataUtfJsonTextWriter` has been created?

Reply:
I tried this and it gives errors.

Reply:
What kind of errors. The `ODataUtf8JsonTextWriter` is designed to write a single value. In that case, its content type should not change during its lifetime because a single value should not have multiple content types.

Reply:
resolved.